client/repo.c

```diff
@@ -691,6 +692,60 @@ TDNFGetRepoMD(
                   pRepoMDRel,
                   &pRepoMD);
     BAIL_ON_TDNF_ERROR(dwError);
+
+    if (pRepoData->pszSnapshotUrl) {
+        time_t now = time(NULL);
+        int needDownload = 0, nIsRemote = 0;
+        struct stat st = {0};
+
+        if (pRepoData->pszSnapshotUrl[0] == '/') {
+            SET_STRING(pRepoData->pszSnapshotFile, pRepoData->pszSnapshotUrl);
+        } else {
+            dwError = TDNFUriIsRemote(pRepoData->pszSnapshotUrl, &nIsRemote);
+            if (dwError) {
+                /* should be a relative path, relative to repo dir */
+                dwError = TDNFJoinPath(
+                              &pRepoData->pszSnapshotFile,
+                              pTdnf->pConf->pszRepoDir,
+                              pRepoData->pszSnapshotUrl,
+                              NULL);
+                BAIL_ON_TDNF_ERROR(dwError);
+            } else {
+                if (nIsRemote) {
+                    /* we need a unique name based on URL locally, so we automatically refresh on config changes
+                       "snapshot" % URL => "snapshot-12345678" */
+                    dwError = SolvCreateRepoCacheName(TDNF_REPO_METADATA_SNAPSHOT, pRepoData->pszSnapshotUrl, &pszSnapshotFile);
+                    BAIL_ON_TDNF_ERROR(dwError);
+
+                    dwError = TDNFGetCachePath(pTdnf, pRepoData,
+                                               pszSnapshotFile, NULL,
+                                               &pRepoData->pszSnapshotFile);
+                    BAIL_ON_TDNF_ERROR(dwError);
+
+                    if (stat(pRepoData->pszSnapshotFile, &st) < 0) {
+                        if (errno == ENOENT)
+                            needDownload = 1;
+                        else {
+                            dwError = errno;
+                            BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+                        }
+                    } else if ((now - st.st_ctime) > pRepoData->lMetadataExpire)
+                        needDownload = 1;
+
+                    if (needDownload) {
+                        dwError = TDNFDownloadFile(pTdnf, pRepoData, pRepoData->pszSnapshotUrl, pRepoData->pszSnapshotFile, pRepoData->pszId);
+                        BAIL_ON_TDNF_ERROR(dwError);
+                    }
+                } else if (strncmp(pRepoData->pszSnapshotUrl, "file://", 7) == 0) {
```

Comment:
JFYI: instead of 7 you can use `sizeof("file://") - 1`

Reply:
I know, but this makes it easier to read, IMHO.